File: solutions/erlang/robot-simulator/2/robot_simulator.erl
-module(robot_simulator).
-behavior(gen_server).

%% Exercism API
-export([ create/0
        , place/3
        , direction/1
        , position/1
        , left/1
        , right/1
        , advance/1
        , control/2
        ]).

%% API
-export([start_link/0]).

%% gen_server callbacks
-export([init/1, handle_call/3, handle_cast/2, handle_info/2, terminate/2, code_change/3]).

-define(SERVER, robot).

-define(DIRECTIONS, {north, east, south, west}).

-record(state, { direction
               , position = {undefined, undefined}
               }).

%%% API

start_link() ->
  gen_server:start_link({local, ?SERVER}, ?MODULE, [], []).

%%% gen_server callbacks

init([]) ->
  {ok, #state{}}.

handle_call(direction, _From, State) ->
  {reply, State#state.direction, State};
handle_call(position, _From, State) ->
  {reply, State#state.position, State};
handle_call({place, Direction, Position}, _From, State) ->
  {reply, ok, State#state{ direction = Direction, position = Position}};
handle_call({turn, Hand}, _From, State) ->
  NewDirection = turn(Hand, State),
  {reply, ok, NewDirection};
handle_call(advance, _From, State) ->
  NewPosition = move(State),
  {reply, ok, NewPosition};
handle_call({control, Command}, _From, State) ->
  NewState = lists:foldl(fun(C, Acc) -> command(C, Acc) end, State, Command),
  {reply, ok, NewState};
handle_call(_Request, _From, State) ->
  {reply, ok, State}.

handle_cast(_Msg, State) -> {noreply, State}.

handle_info(_Info, State) -> {noreply, State}.

terminate(_Reason, _State) -> ok.

code_change(_OldVsn, State, _Extra) -> {ok, State}.

%%% Exercism API

create() ->
  % NOTE: create_tests will fail if it isn't the first run.
  case gen_server:start_link({local, ?SERVER}, ?MODULE, [], []) of
    {ok, _} -> ?SERVER;
    {error, {already_started, _}} -> ?SERVER
  end.

place(robot, Direction, Position) ->
  gen_server:call(robot, {place, Direction, Position}).

direction(robot) ->
  gen_server:call(robot, direction).

position(robot) ->
  gen_server:call(robot, position).

left(robot) ->
  gen_server:call(robot, {turn, left}).

right(robot) ->
  gen_server:call(robot, {turn, right}).

advance(robot) ->
  gen_server:call(robot, advance).

control(robot, "unknown") ->
  ok;
control(robot, Command) ->
  gen_server:call(robot, {control, Command}).

%%% Internal functions

turn(left, State) when State#state.direction =:= north ->
  State#state{direction = west};
turn(left, State) when State#state.direction =:= west ->
  State#state{direction = south};
turn(left, State) when State#state.direction =:= south ->
  State#state{direction = east};
turn(left, State) when State#state.direction =:= east ->
  State#state{direction = north};
turn(right, State) when State#state.direction =:= north ->
  State#state{direction = east};
turn(right, State) when State#state.direction =:= west ->
  State#state{direction = north};
turn(right, State) when State#state.direction =:= south ->
  State#state{direction = west};
turn(right, State) when State#state.direction =:= east ->
  State#state{direction = south}.

move(State) when State#state.direction =:= north ->
  {X, Y} = State#state.position,
  State#state{position = {X, Y + 1}};
move(State) when State#state.direction =:= east ->
  {X, Y} = State#state.position,
  State#state{position = {X + 1, Y}};
move(State) when State#state.direction =:= south ->
  {X, Y} = State#state.position,
  State#state{position = {X, Y - 1}};
move(State) when State#state.direction =:= west ->
  {X, Y} = State#state.position,
  State#state{position = {X - 1, Y}}.

command($R, State) ->
  turn(right, State);
command($L, State) ->
  turn(left, State);
command($A, State) ->
  move(State).

File: solutions/elixir/robot-simulator/2/lib/robot_simulator.ex
defmodule RobotSimulator do
  @type direction :: :north | :east | :south | :west
  @type position :: {integer, integer}
  @type robot :: {direction, position}

  @directions [:north, :east, :south, :west]

  defguardp is_direction(direction) when direction in @directions
  defguardp is_position(x, y) when is_integer(x) and is_integer(y)

  @doc """
  Create a Robot Simulator given an initial direction and position.
  Valid directions are: `:north`, `:east`, `:south`, `:west`
  """
  @spec create(direction :: atom, position :: position) :: robot
  def create(direction \\ :north, position \\ {0, 0})
  def create(direction, _position) when not is_direction(direction) do
    {:error, "invalid direction"}
  end
  def create(direction, {x, y}) when is_position(x, y), do: {direction, {x, y}}
  def create(_direction, _position), do: {:error, "invalid position"}

  @doc """
  Simulate the robot's movement given a string of instructions.
  Valid instructions are: "R" (turn right), "L", (turn left), and "A" (advance)
  """
  @spec simulate(robot :: robot, instructions :: charlist()) :: robot
  def simulate(robot, ""), do: robot
  def simulate(robot, <<?A, rest::binary>>), do: simulate(advance(robot), rest)
  def simulate(robot, <<?L, rest::binary>>), do: simulate(rotate(?L, robot), rest)
  def simulate(robot, <<?R, rest::binary>>), do: simulate(rotate(?R, robot), rest)
  def simulate(_robot, _invalid), do: {:error, "invalid instruction"}

  @doc """
  Return the robot's direction.
  Valid directions are: `:north`, `:east`, `:south`, `:west`
  """
  @spec direction(robot :: robot) :: direction
  def direction({direction, _}), do: direction

  @doc """
  Return the robot's position.
  """
  @spec position(robot :: robot) :: position
  def position({_, position}), do: position

  @spec rotate(rotation :: char, robot :: robot) :: robot
  defp rotate(?R, {:north, position}), do: {:east, position}
  defp rotate(?R, {:east, position}), do: {:south, position}
  defp rotate(?R, {:south, position}), do: {:west, position}
  defp rotate(?R, {:west, position}), do: {:north, position}
  defp rotate(?L, {:north, position}), do: {:west, position}
  defp rotate(?L, {:east, position}), do: {:north, position}
  defp rotate(?L, {:south, position}), do: {:east, position}
  defp rotate(?L, {:west, position}), do: {:south, position}

  @spec advance(robot :: robot) :: robot
  defp advance({:north, {x, y}}), do: {:north, {x, y + 1}}
  defp advance({:east, {x, y}}), do: {:east, {x + 1, y}}
  defp advance({:south, {x, y}}), do: {:south, {x, y - 1}}
  defp advance({:west, {x, y}}), do: {:west, {x - 1, y}}
end
